pkg/controller/istio/istio_controller.go

```diff
@@ -0,0 +1,465 @@
+	reqLogger := log.WithValues("Request.Namespace", request.Namespace, "Request.Name", request.Name)
+	reqLogger.V(1).Info("Reconciling Istio")
+
+	if request.NamespacedName != utils.DefaultInstanceKey {
+		reqLogger.V(1).Info("Istio resource named %q is not recognised, name it %q",
+			request.Name, utils.DefaultInstanceKey.Name)
+		return reconcile.Result{}, nil
+	}
+
+	// Get the Istio CR.
+	instance := &operatorv1.Istio{}
+	err := r.Get(ctx, utils.DefaultInstanceKey, instance)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			reqLogger.Info("Istio object not found")
+			r.status.OnCRNotFound()
+			return reconcile.Result{}, nil
+		}
+		r.status.SetDegraded(operatorv1.ResourceReadError,
+			fmt.Sprintf("Error querying for Istio CR: failed to get Istio %q", utils.DefaultInstanceKey),
+			err, reqLogger)
+		return reconcile.Result{}, err
+	}
+	updateDefaults(instance)
+	r.status.OnCRFound()
+
+	// SetMetaData in the TigeraStatus such as observedGenerations.
+	defer r.status.SetMetaData(&instance.ObjectMeta)
+
+	defer r.setCondition(ctx, instance, reqLogger)
+
+	setCurrentCondition(instance, v1.ComponentProgressing, reasonInfo, "Deploying Tigera Istio resources")
+
+	if res, err, finished := r.maintainFinalizer(ctx, instance, reqLogger); err != nil || finished {
+		return res, err
+	}
+
+	// Get the Installation, for k8s provider info.
+	variant, installation, err := utils.GetInstallation(ctx, r)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			r.status.SetDegraded(operatorv1.ResourceNotFound, "Installation not found", err, reqLogger)
+			return reconcile.Result{}, nil
+		}
+		r.status.SetDegraded(operatorv1.ResourceReadError, "Error querying installation", err, reqLogger)
+		setCurrentCondition(instance, v1.ComponentDegraded, reasonErr, "Installation resource not found")
+		return reconcile.Result{}, err
+	}
+
+	if variant == "" {
+		r.status.SetDegraded(operatorv1.ResourceNotReady, "Waiting for Installation Variant to be set", nil, reqLogger)
+		return reconcile.Result{Requeue: true}, nil
+	}
+
+	// Get the Kubernetes Gateway API CRDs.
+	essentialCRDs, optionalCRDs := gatewayapi.K8SGatewayAPICRDs(installation.KubernetesProvider)
+
+	// Check CRDs are present and only create it if not
+	handler := utils.NewComponentHandler(log, r, r.scheme, nil)
+	handler.SetCreateOnly()
+	err = handler.CreateOrUpdateOrDelete(ctx, render.NewPassthrough(essentialCRDs...), nil)
+	if err != nil && !errors.IsAlreadyExists(err) {
+		r.status.SetDegraded(operatorv1.ResourceCreateError, "Error creating gateway API CRDs", err, log)
+		setCurrentCondition(instance, v1.ComponentDegraded, reasonErr, fmt.Sprintf("Error creating Gateway API resources: %s", err.Error()))
+		return reconcile.Result{}, err
+	}
+	err = handler.CreateOrUpdateOrDelete(ctx, render.NewPassthrough(optionalCRDs...), nil)
+	if err != nil && !errors.IsAlreadyExists(err) {
+		reqLogger.Info("Could not render all optional gateway API CRDs", "err", err)
```

Comment:
This could be quite spammy - we shouldn't log anything at regular Info level in the reconcile loop.

Reply:
When do we expect to hit this branch?

Reply:
Yes, indeed. By default OCP environments blocks to install some CRDs, that here they were split into optional ones, once they are really not required for the feature to work.

Reply:
How should we inform so that the optional ones had failed?